### PR TITLE
Added Basic Auth Support

### PR DESCRIPTION
**Pull request by reifl**
fixes https://github.com/evcc-io/app/issues/5

Added support for Basic Auth. When the Server is found via mDns it submits the Address to the ServerManualScreen where the User can input the Credentials.

### CODE REVIEW

**Comment by reifl** [comment]
i've accidentally commited an update of expo-dev-client and react-native-v8 with my initial pull request. Needed to update them to build for android, the "npm run android" always exited with an error but thats another problem and shouldn't be in this pull request so i commited a new version of the package.json where i've reverted to the original package version.

**Comment by naltatis** [comment]
@reifl tried to test it but got errors:

Bildschirmfoto 2024-08-18 um 19 19 38

I also just updated the npm deps. Maybe this solves your issues above.

We should also support the auto-discovery use-case. Maybe it's a good idea to check auto-detect if auth is required during the `validateAndSaveURL` and then prompt the user for credentials if we dont have them.

**Comment by reifl** [comment]
I've merged your update to my fork and tried it. It works on my Android Device, unluckly i don't have a mac on hand to test the iOS build. I will try to organize one next week so i can check whats happening.

To your auto-discovery use-case. you mean the "Suche starten" on the Main Screen? If so this use-case should be covered. When the user selects a server which needs basic auth (401 Status code) it switches to the "Adresse manuell eingeben" mask with prefilled Server Information.

**Comment by naltatis** [comment]
> If so this use-case should be covered. When the user selects a server which needs basic auth (401 Status code) it switches to the "Adresse manuell eingeben" mask with prefilled Server Information.

Ah ok, perfect

**Comment by naltatis** [comment]
> I've merged your update to my fork and tried it.

Tried it after the upgrade but still not able to run in iOS simulator or on-device. I'll try to run it on my android test device during the week.

Another thing: looks like your code is not prettier-formatted. Please `npm run format` or enable format-on-save in your editor.

**Comment by reifl** [comment]
iOS version should now work and ran "npm run format"

File: interfaces/basic-auth-information.tsx [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export interface BasicAuthInformation {
+  basicAuthRequired: boolean;
```
Comment (naltatis) [inline interfaces/basic-auth-information.tsx:2]:
```suggestion
  required: boolean;
```

Should be unique enough in this context

File: components/AppContext.tsx [hunk 2]
```diff
@@ -1,10 +1,15 @@
 import React, { createContext, useContext, useState, useEffect } from "react";
 import AsyncStorage from "@react-native-async-storage/async-storage";
+import { BasicAuthInformation } from "../interfaces/basic-auth-information";
 
 // Create a context
 const AppContext = createContext({
   serverUrl: "",
-  updateServerUrl: async (url: string) => {
+  basicAuthInformation: { basicAuthRequired: false } as BasicAuthInformation,
```
Comment (naltatis) [inline components/AppContext.tsx:8]:
```suggestion
  basicAuth: { required: false } as BasicAuth,
```

would keep it shorter

File: components/ServerForm.tsx [hunk 3]
```diff
@@ -55,10 +85,58 @@ export default function ServerForm({
...
+        checked={basicAuth.basicAuthRequired}
+        onChange={(nextValue) => setBasicAuthRequired(nextValue)}
+      >
+        Anmeldung erforderlich
+      </CheckBox>
+      {basicAuth.basicAuthRequired && (
+        <Input
+          style={{ marginTop: 5 }}
+          size="large"
+          status={validateUsername() ? "basic" : "danger"}
```
Comment (naltatis) [inline components/ServerForm.tsx:106]: Having it initially read is irritating. I'd skip validation completely on both fields and always do basic status without label.

File: components/ServerForm.tsx [hunk 4]
```diff
@@ -55,10 +85,58 @@ export default function ServerForm({
...
+          autoCorrect={false}
+          placeholder="Benutzer"
+          label="Benutzer"
+          ref={usernameRef}
+          onSubmitEditing={() => passwordRef.current.focus()}
+        />
+      )}
+      {basicAuth.basicAuthRequired && (
+        <Input
+          style={{ marginTop: 5 }}
```
Comment (naltatis) [inline components/ServerForm.tsx:122]: I've adjusted the spaced a bit locally. We have enough vertical space to make this less condense.

Simulator Screenshot - iPhone 15 - 2024-08-19 at 20 53 40

Simulator Screenshot - iPhone 15 - 2024-08-19 at 20 53 43

**Comment by naltatis** [comment]
@reifl I accidentally merged your PR to main. Not quite sure how this happened `gh pr` magic. But I'll then make the above changes directly in main. Thanks for your work 💚

**Pull request merged**